src/lib/Hydra/Helper/AddBuilds.pm

```diff
@@ -105,10 +105,14 @@ sub fetchInputBuild {
     my $relName = ($prevBuild->releasename or $prevBuild->nixname);
     my $version = $2 if $relName =~ /^($pkgNameRE)-($versionRE)$/;
 
+    my $mainOutput = getMainOutput($prevBuild);
+
     return
-        { storePath => getMainOutput($prevBuild)->path
+        { storePath => $mainOutput->path
         , id => $prevBuild->id
         , version => $version
+        , outputName => $mainOutput->name
+        , drvPath => $prevBuild->drvPath
```

Comment:
On second thought, the derivations *are* kept if `gc-keep-derivations` is set in `nix.conf`, which happens to be the default. So this should be fine. Maybe you can make passing `drvPath` conditional on `isValidPath($prevBuild->drvPath)`.